lib/broadway/options.ex

```diff
@@ -218,10 +218,27 @@ defmodule Broadway.Options do
               """
             ],
             batch_size: [
-              type: :pos_integer,
+              type: {:or, [:pos_integer, {:fun, 2}]},
               default: 100,
               doc: """
-              The size of the generated batches.
+              The size of the generated batches. Default value is `100`.
+              It can also be a function that takes a
+              `Broadway.Message` and last calculated value returned from this function.
+              Value of the second parameter is `nil` on first call for one batch and you should set initial counter.
+              The function must return atom `:done` if passed in message indicates a full batch stop.
+              Default batch size is actually converted to this:
+              `fn _message, remained ->
+                remained = if(remained == nil, do: batch_size - 1, else: remained - 1)
+                if(remained == 0, do: :done, else: remained)
```

Comment:
I am wondering if we should instead have an API like this:

```elixir
{:emit, acc} | {:cont, acc}
```

It will give the batcher more control over the accumulator.

Reply:
The return type makes it more clear than a 0.  But if the batcher received a {:emit, acc}, the acc should be reset to what the user set in options, right?

Reply:
The function should always have the control over the accumulator.